File: python/calculator/calculator.py
from playwright.sync_api import sync_playwright
import csv, sys

CALCULATOR_URL = "https://treasurydirect.gov/BC/SBCPrice"


def calculate_bond(page, row):
    page.locator("[name='Series']").select_option(row[0])
    page.locator("[name='Denomination']").select_option(row[1])
    page.locator("[name='IssueDate']").fill(row[2])
    page.locator("[name='SerialNumber']").fill(row[3])
    page.locator("[value='CALCULATE']").click()


if not sys.argv[1]:
    print(
        "Error: Provide a CSV containing Series, Denomination, Issue Date (MMYYYY), Serial Number (optional)"
    )
    sys.exit(1)
input_filepath = sys.argv[1]

with sync_playwright() as p:
    browser = p.firefox.launch(headless=False)
    page = browser.new_page()
    page.goto(CALCULATOR_URL)
    with open(input_filepath) as fp:
        csv_reader = csv.reader(fp)
        for row in csv_reader:
            calculate_bond(page, row)
    page.get_by_alt_text("Save this list.").click()
    html = page.content()
    with open("saved-bonds.html", "w") as file:
        file.write(html)

File: playwright/javascript/src/calculator.js
import fs from "fs";
import { performance } from "perf_hooks";
import papa from "papaparse";
import { firefox } from "playwright";

const CALCULATOR_URL = "https://treasurydirect.gov/BC/SBCPrice";

const readCSVFile = async (filePath) => {
  return new Promise((resolve, reject) => {
    const fileStream = fs.createReadStream(filePath, "utf8");

    const data = [];

    papa.parse(fileStream, {
      header: false,
      skipEmptyLines: true,
      step: (row) => {
        data.push(row.data);
      },
      complete: () => {
        resolve(data);
      },
      error: (error) => {
        reject(error);
      },
    });
  });
};

const calculateBond = async (page, bond) => {
  await page.locator("[name='Series']").selectOption(bond[0]);
  await page.locator("[name='Denomination']").selectOption(bond[1]);
  await page.locator("[name='IssueDate']").fill(bond[2]);
  await page.locator("[name='SerialNumber']").fill(bond[3]);
  await page.locator("[value='CALCULATE']").click();
};

(async () => {
  const args = process.argv.slice(2);
  if (args.length < 1) {
    console.error("Please provide the path to a CSV file.");
    process.exit(1);
  }

  const csvFilePath = args[0];
  const bonds = await readCSVFile(csvFilePath);
  const browser = await firefox.launch({
    headless: false,
  });
  const context = await browser.newContext();
  const page = await context.newPage();
  await page.goto("https://treasurydirect.gov/BC/SBCPrice");

  const startTime = performance.now();
  for (const bond of bonds) {
    await calculateBond(page, bond);
  }
  await page.getByAltText("Save this list.").click();
  const html = await page.content();
  const endTime = performance.now();
  console.log(
    `Call to calculate bonds and get page content took ${
      endTime - startTime
    } milliseconds`
  );
  fs.writeFile("saved-bonds.html", html, "utf8", (err) => {
    if (err) {
      console.error(`Error writing to file:`, err);
    } else {
      console.log(`File written successfully to file`);
    }
  });
  await context.close();
  await browser.close();
})();
